### PR TITLE
add futures

### PR DESCRIPTION
**Pull request by thde**
implements #43

### CODE REVIEW

File: src/main/kotlin/ch/hsr/dsl/dwrtc/signaling/ClientService.kt [hunk 1]
```diff
@@ -43,34 +49,39 @@ class ClientService constructor(peerPort: Int? = findFreePort()) {
...
 
-        val peerIdGet = peer.get(Number160.createHash(sessionId)).start().awaitUninterruptibly()
-        return if (peerIdGet.isSuccess && peerIdGet.data() != null) {
-            logger.info { "found client" }
+        val dhtFuture = peer.get(Number160.createHash(sessionId)).start()
+        val future = GetCustomFuture<IExternalClient, PeerAddress>(
+            dhtFuture
+        ) { peerAddress -> ExternalClient(sessionId, peerAddress, peer) }
+
+        future.onFailure { "find client with $sessionId failed" }
```
Comment (thde) [inline src/main/kotlin/ch/hsr/dsl/dwrtc/signaling/ClientService.kt]: `logger.info`

File: src/main/kotlin/ch/hsr/dsl/dwrtc/signaling/ClientService.kt [hunk 2]
```diff
@@ -43,34 +49,39 @@ class ClientService constructor(peerPort: Int? = findFreePort()) {
...
 
-    fun findClient(sessionId: String): ExternalClient {
+    override fun findClient(sessionId: String): GetFuture<IExternalClient> {
         logger.info { "try to find client $sessionId" }
 
-        val peerIdGet = peer.get(Number160.createHash(sessionId)).start().awaitUninterruptibly()
-        return if (peerIdGet.isSuccess && peerIdGet.data() != null) {
-            logger.info { "found client" }
+        val dhtFuture = peer.get(Number160.createHash(sessionId)).start()
+        val future = GetCustomFuture<IExternalClient, PeerAddress>(
```
Comment (lroellin) [inline src/main/kotlin/ch/hsr/dsl/dwrtc/signaling/ClientService.kt]: Maybe we can format this a bit nicer

**Comment by lroellin** [comment]
Fancy-pantsy!

File: src/main/kotlin/ch/hsr/dsl/dwrtc/util/TomP2P.kt [hunk 3]
```diff
@@ -19,3 +20,27 @@ fun BaseFuture.onSuccess(emitter: () -> Unit) {
 fun BaseFuture.onFailure(emitter: (failedReason: String) -> Unit) {
     this.onComplete { result -> if (result.isFailed) emitter(result.failedReason()) }
 }
+
+fun <T, U> FutureGet.onGetCustom(emitter: (data: U?, future: Future) -> Unit, transformer: (T) -> U) {
+    this.addListener(object : BaseFutureAdapter<FutureGet>() {
+        override fun operationComplete(future: FutureGet) {
+            val data = transformer(future.data()?.`object`() as T) ?: null
```
Comment (lroellin) [inline src/main/kotlin/ch/hsr/dsl/dwrtc/util/TomP2P.kt]: IntelliJ gives me a notice on compilation:
Warning:(27, 68) Kotlin: Right operand of elvis operator (?:) is useless if it is null

**Pull request merged**